### PR TITLE
ncurses: migrate to Conan v2

### LINKED ISSUE #16627: [package] ncurses/6.3: Cannot Import Package for Conan 2.0
### Description

```
*********************************************************
Recipe 'ncurses/6.3' seems broken.
It is possible that this recipe is not Conan 2.0 ready
If the recipe comes from ConanCenter check: https://conan.io/cci-v2.html
If it is your recipe, check if it is updated to 2.0
*********************************************************

ERROR: Package 'ncurses/6.3' not resolved: ncurses/6.3: Cannot load recipe.
Error loading conanfile at '/home/fortem/.conan2/p/ncurse90fb70e239f8/e/conanfile.py': Unable to load conanfile in /home/fortem/.conan2/p/ncurse90fb70e239f8/e/conanfile.py
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/fortem/.conan2/p/ncurse90fb70e239f8/e/conanfile.py", line 1, in <module>
    from conans import AutoToolsBuildEnvironment, ConanFile, tools
ImportError: cannot import name 'AutoToolsBuildEnvironment' from 'conans' (/home/fortem/Development/dronehunter-dev/venv/lib/python3.8/site-packages/conans/__init__.py)

```

### Package and Environment Details

* Package Name/Version: **ncurses/6.3**
* Operating System+version: **Linux Ubuntu 20.04**
* Compiler+version: **GCC 9**
* Docker image: **conanio/gcc9**
* Conan version: **conan 2.0.2**
* Python version: **Python 3.7.4**

### Conan profile

[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=9
os=Linux

### Steps to reproduce

conan install .

### Logs

<details><summary>Click to expand log</summary>

```
======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Debug
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=9
os=Linux

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu17
compiler.libcxx=libstdc++11
compiler.version=9
os=Linux

======== Computing dependency graph ========
behaviortree.cpp/4.0.1: Not found in local cache, looking in remotes...
behaviortree.cpp/4.0.1: Checking remote: conancenter
Decompressing conan_export.tgz
behaviortree.cpp/4.0.1: Downloaded recipe revision 1dc19e508c23f27ed830a96f42db3643
Graph root
    conanfile.py (test/1.0.0): /home/fortem/Development/test/conanfile.py
Requirements
    behaviortree.cpp/4.0.1#1dc19e508c23f27ed830a96f42db3643 - Downloaded (conancenter)
*********************************************************

*********************************************************
Recipe 'ncurses/6.3' seems broken.
It is possible that this recipe is not Conan 2.0 ready
If the recipe comes from ConanCenter check: https://conan.io/cci-v2.html
If it is your recipe, check if it is updated to 2.0
*********************************************************

ERROR: Package 'ncurses/6.3' not resolved: ncurses/6.3: Cannot load recipe.
Error loading conanfile at '/home/fortem/.conan2/p/ncurse90fb70e239f8/e/conanfile.py': Unable to load conanfile in /home/fortem/.conan2/p/ncurse90fb70e239f8/e/conanfile.py
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/fortem/.conan2/p/ncurse90fb70e239f8/e/conanfile.py", line 1, in <module>
    from conans import AutoToolsBuildEnvironment, ConanFile, tools
ImportError: cannot import name 'AutoToolsBuildEnvironment' from 'conans' (/home/fortem/Development/dronehunter-dev/venv/lib/python3.8/site-packages/conans/__init__.py)


```

</details>

### LINKED ISSUE #20084: [package] ncurses/6.4: Seems to not be conan2 compatible
### Description

Was told this package may not have been updated to conan/2 as i was updating my own project

### Package and Environment Details

* Package Name/Version: **ncurses/6.4**
* Operating System+version: **Linux Ubuntu 20.04 (win10 wsl2)**
* Compiler+version: **GCC 9.4**
* Conan version: **conan 2.0.10**
* Python version: **Python 3.8.10**

### Conan profile

Host profile:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu14
compiler.libcxx=libstdc++11
compiler.version=9
os=Linux

Build profile:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.cppstd=gnu14
compiler.libcxx=libstdc++11
compiler.version=9
os=Linux

### Steps to reproduce
git  clone https://github.com/benny-edlund/task-pool/
cd task-pool/examples
git checkout conan_examples_update
conan install . --output-folder=build --build=missing --settings=build_type=Release

### Logs

<details><summary>Click to expand log</summary>

```
*********************************************************
Recipe 'ncurses/6.3' seems broken.
It is possible that this recipe is not Conan 2.0 ready
If the recipe comes from ConanCenter, report it at https://github.com/conan-io/conan-center-index/issues
If it is your recipe, check if it is updated to 2.0
*********************************************************

ERROR: Package 'ncurses/6.3' not resolved: ncurses/6.3: Cannot load recipe.
Error loading conanfile at '/home/benny/.conan2/p/ncurse90fb70e239f8/e/conanfile.py': Unable to load conanfile in /home/benny/.conan2/p/ncurse90fb70e239f8/e/conanfile.py
  File "<frozen importlib._bootstrap_external>", line 848, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/benny/.conan2/p/ncurse90fb70e239f8/e/conanfile.py", line 1, in <module>
    from conans import AutoToolsBuildEnvironment, ConanFile, tools
ImportError: cannot import name 'AutoToolsBuildEnvironment' from 'conans' (/home/benny/.local/lib/python3.8/site-packages/conans/__init__.py)```

</details>

### PR DESCRIPTION
**Pull request by valgur**
Some notable changes:
* Moved the definition of the compatibility CMake module to an external file.
* Condensed the component definitions in `package_info()`.
* Added an unofficial `Curses::Curses` CMake target. CMake currently does not set one (https://gitlab.kitware.com/cmake/cmake/-/issues/23051), but I think having it is helpful and much less error-prone than using package variables. 

These can probably be closed if this PR gets merged:
* Resolves #17805
* Resolves #18568
* Resolves #16627
* Resolves #20084

### CODE REVIEW

File: recipes/ncurses/all/conanfile.py [hunk 1]
```diff
@@ -134,108 +138,116 @@ def _configure_autotools(self):
...
+        os.unlink(os.path.join(self.package_folder, "bin", f"ncurses{self._suffix}{self._major_version}-config"))
+        copy(self, "*.cmake",
+             src=os.path.join(self.export_sources_folder, "cmake"),
+             dst=os.path.join(self.package_folder, self._module_subfolder))
+        fix_apple_shared_install_name(self)
 
     @property
     def _suffix(self):
         res = ""
+        # https://github.com/mirror/ncurses/blob/v6.4/configure.in#L1393
```
Comment (uilianries) [inline recipes/ncurses/all/conanfile.py:247]: nice, thank you for pointing it.

File: recipes/ncurses/all/conanfile.py [hunk 2]
```diff
@@ -134,108 +138,116 @@ def _configure_autotools(self):
...
+        autotools = Autotools(self)
+        autotools.configure()
+        autotools.make()
 
     @property
     def _major_version(self):
-        return tools.Version(self.version).major
-
-    @staticmethod
-    def _create_cmake_module_alias_targets(module_file):
```
Comment (uilianries) [inline recipes/ncurses/all/conanfile.py:205]: I have to agree, a separated cmake file looks better for this case.

File: recipes/ncurses/all/conanfile.py [hunk 3]
```diff
@@ -41,90 +49,86 @@ class NCursesConan(ConanFile):
...
+        export_conandata_patches(self)
+        copy(self, "*.cmake", src=self.recipe_folder, dst=self.export_sources_folder)
 
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
+        if self.options.with_ticlib == "auto":
+            self.options.with_ticlib = self.settings.os != "Windows"
+        if self.options.with_tinfo == "auto":
+            self.options.with_tinfo = self.settings.os != "Windows"
```
Comment (uilianries) [inline recipes/ncurses/all/conanfile.py:70]:
```suggestion
        if self.options.with_ticlib == "auto":
            self.options.with_ticlib = self.settings.os != "Windows"
        if self.options.with_tinfo == "auto":
            self.options.with_tinfo = self.settings.os != "Windows"
```

Need to move to configure() method. config options can not acquire options values passed by users.

Comment (valgur) [reply recipes/ncurses/all/conanfile.py:70]: Hmm, I was getting my hopes up for this working based on https://github.com/conan-io/conan-center-index/blob/71b82701d768a332ea97a9874a1fd01942ec1440/recipes/arrow/all/conanfile.py#L161-L175
Tested it again locally just to be sure and it indeed does not work. The linked logic in arrow is totally broken then, I guess?

File: recipes/ncurses/all/conanfile.py [hunk 4]
```diff
@@ -1,31 +1,39 @@
...
         "fPIC": [True, False],
         "with_widec": [True, False],
         "with_extended_colors": [True, False],
         "with_cxx": [True, False],
         "with_progs": [True, False],
-        "with_ticlib": ["auto", True, False],
+        "with_ticlib": [True, False],
         "with_reentrant": [True, False],
-        "with_tinfo": ["auto", True, False],
+        "with_tinfo": [True, False],
```
Comment (uilianries) [inline recipes/ncurses/all/conanfile.py:36]: Please, inform how do you change these values in term of always True, but considering that the system will actually do a check to enable/disable first.

Comment (valgur) [reply recipes/ncurses/all/conanfile.py]: I restored the "auto" default value for clarity and backwards compatibility.

File: recipes/ncurses/all/conanfile.py [hunk 5]
```diff
@@ -134,108 +137,116 @@ def _configure_autotools(self):
...
+            ]
+        if is_msvc(self):
+            build = host = f"{self.settings.arch}-w64-mingw32-msvc"
+            tc.configure_args += [
                 "ac_cv_func_getopt=yes",
                 "ac_cv_func_setvbuf_reversed=no",
-            ])
-            autotools.cxx_flags.append("-EHsc")
-            if tools.Version(self.settings.compiler.version) >= 12:
-                autotools.flags.append("-FS")
```
Comment (Ahajha) [inline recipes/ncurses/all/conanfile.py:159]: Re-adding this flag like this fixed the build for me locally (6.2, Debug, MDd, shared):
```python
if Version(self.settings.compiler.version) >= 12:
    tc.extra_cxxflags.append("-FS")
```

Comment (valgur) [reply recipes/ncurses/all/conanfile.py:159]: Thanks, @Ahajha! Looks like this fixed the issue.

File: recipes/ncurses/all/cmake/conan-official-ncurses-targets.cmake
Comment (uilianries) [inline recipes/ncurses/all/cmake/conan-official-ncurses-targets.cmake:1]: Okay, I checked https://cmake.org/cmake/help/latest/module/FindCurses.html and this file follows the variables there.

**Comment by RubenRBS** [comment]
I've tested locally with these combinations which seem to work:

```
10  conan create . --version=6.3 -o "&:with_pcre2=True" --build=missing -c:h "tools.files.download:verify=False" -c:b "tools.files.download:verify=False" -o "*/*:shared=True"
   11  conan create . --version=6.3 -o "&:with_pcre2=True" --build=missing -c:h "tools.files.download:verify=False" -c:b "tools.files.download:verify=False" -o "*/*:shared=True" -o "&:with_reentrant=True"
   12  conan create . --version=6.3 -o "&:with_pcre2=True" --build=missing -c:h "tools.files.download:verify=False" -c:b "tools.files.download:verify=False" -o "*/*:shared=True" -o "&:with_reentrant=True" -o "&:with_progs=False"
```

File: recipes/ncurses/all/conanfile.py [hunk 6]
```diff
@@ -41,90 +49,85 @@ class NCursesConan(ConanFile):
...
-            return self.options.with_tinfo
+    def export_sources(self):
+        export_conandata_patches(self)
+        copy(self, "*.cmake", src=self.recipe_folder, dst=self.export_sources_folder)
 
     def config_options(self):
         if self.settings.os == "Windows":
             del self.options.fPIC
+        # Set the default value based on OS
+        self.options.with_ticlib = self.settings.os != "Windows"
```
Comment (RubenRBS) [inline recipes/ncurses/all/conanfile.py:68]: This is overriding whatever the user is setting these options as even if they are not set to auto, right?

**Review by RubenRBS — COMMENTED** [review]
Question about the handling of with_ticlib, looks good otherwise :)

File: recipes/ncurses/all/conanfile.py (hunk 6 above)
Comment (RubenRBS) [reply recipes/ncurses/all/conanfile.py:68]: I see https://github.com/conan-io/conan-center-index/pull/20355#discussion_r1397324700 but not clear to me what the correct way would be

Comment (valgur) [reply recipes/ncurses/all/conanfile.py:68]: This only overwrites the default "auto" value with a correct OS-specific default value. User-specified values for these options are not affected.

Believe me, this is approach is positively fantastic and I'm glad it works. Any alternatives approaches would need an insane amount of hacking around `self.options` and `self.info.options` being expected in different methods.

**Comment by uilianries** [comment]
@valgur Could you please add `test_v1_package` too. It's missing and this recipe is not a new one.

**Review by uilianries — DISMISSED** [review]
LGTM

**Review by RubenRBS — APPROVED** [review]

**Review by samuel-emrys — APPROVED** [review]

**Pull request merged**